quic/s2n-quic-tests/src/tests/drop_initial_packets.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+use s2n_quic::provider::{
+    endpoint_limits::{ConnectionAttempt, Limiter, Outcome},
+    tls::default::{self as tls},
+};
+use s2n_quic_core::{connection::error::Error, endpoint};
+
+/// A custom limiter that always returns Outcome::close()
+struct AlwaysCloseLimiter;
+
+impl Limiter for AlwaysCloseLimiter {
+    fn on_connection_attempt(&mut self, _info: &ConnectionAttempt) -> Outcome {
+        Outcome::close()
+    }
+}
+
+// This test verifies that when the server would send a CONNECTION_CLOSE frame with
+// error code CONNECTION_REFUSED when the server's limiter returns Outcome::close().
+#[test]
+fn drop_initial_packets_test() {
+    let model = Model::default();
+
+    let connection_close_subscriber = recorder::ConnectionClosed::new();
+    let connection_close_event = connection_close_subscriber.events();
+
+    test(model, |handle| {
+        let server_tls_setup = tls::Server::builder()
+            .with_application_protocols(["h3"].iter())
+            .unwrap()
+            .with_certificate(certificates::CERT_PEM, certificates::KEY_PEM)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let server = Server::builder()
+            .with_io(handle.builder().build()?)?
+            .with_tls(server_tls_setup)?
+            .with_event(tracing_events())?
+            .with_random(Random::with_seed(456))?
+            .with_endpoint_limits(AlwaysCloseLimiter)?
+            .start()?;
+
+        let server_addr = start_server(server)?;
+
+        let client_tls_setup = tls::Client::builder()
+            .with_application_protocols(["h3"].iter())
+            .unwrap()
+            .with_certificate(certificates::CERT_PEM)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let client = Client::builder()
+            .with_io(handle.builder().build()?)?
+            .with_tls(client_tls_setup)?
+            .with_event((tracing_events(), connection_close_subscriber))?
+            .with_random(Random::with_seed(456))?
+            .start()?;
+
+        primary::spawn(async move {
+            let connect = Connect::new(server_addr).with_server_name("localhost");
+
+            // The server should immediately close the connection, so that the handshake will fail
+            client.connect(connect).await.unwrap_err();
```

Comment:
can you make any assertions on the error the client received?

Reply:
And it might be nice for the test to have a positive case -- e.g., by changing AlwaysClose to close after one successful connection. That way we have higher confidence that the rejected connection is rejected for the right reason, rather than e.g. test setup issues.